File: android/rctmgl/src/main/java-v10/com/mapbox/rctmgl/components/location/LocationComponentManager.java
package com.mapbox.rctmgl.components.location;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Color;
import android.graphics.ColorFilter;
import android.graphics.LightingColorFilter;
import android.graphics.drawable.Drawable;

import com.mapbox.geojson.Point;
import com.mapbox.maps.CameraOptions;
import com.mapbox.maps.MapboxMap;
import com.mapbox.maps.Style;
import com.mapbox.maps.plugin.LocationPuck;
import com.mapbox.maps.plugin.LocationPuck2D;
import com.mapbox.maps.plugin.gestures.GesturesUtils;
import com.mapbox.maps.plugin.locationcomponent.LocationComponentPlugin;
import com.mapbox.maps.plugin.locationcomponent.LocationComponentUtils;
import com.mapbox.maps.plugin.locationcomponent.OnIndicatorBearingChangedListener;
import com.mapbox.maps.plugin.locationcomponent.OnIndicatorPositionChangedListener;
import com.mapbox.maps.plugin.locationcomponent.generated.LocationComponentSettings;
import com.mapbox.rctmgl.R;
import com.mapbox.rctmgl.components.location.CameraMode;
import com.mapbox.rctmgl.components.mapview.RCTMGLMapView;
import com.mapbox.rctmgl.location.LocationManager;
import com.mapbox.rctmgl.utils.Logger;

import androidx.annotation.NonNull;
import androidx.appcompat.content.res.AppCompatResources;

/**
 * The LocationComponent on android implements both location tracking and display of user's current location.
 * LocationComponentManager attempts to separate that, so that Camera can ask for location tracking independent of display of user current location.
 * And NativeUserLocation can ask for display of user's current location - independent of Camera's user tracking.
 */
public class LocationComponentManager {
    private RCTMGLMapView mMapView = null;
    private MapboxMap mMap = null;

    private LocationManager mLocationManager = null;
    private LocationComponentPlugin mLocationComponent = null;
    private Context mContext = null;

    private int mCameraMode = CameraMode.NONE;

    private @RenderMode.Mode int mRenderMode = RenderMode.COMPASS;

    private OnIndicatorBearingChangedListener mLocationBearingChangedListener = new OnIndicatorBearingChangedListener() {
        @Override
        public void onIndicatorBearingChanged(double v) {
            if (mFollowUserLocation) {
                mMapView.getMapboxMap().setCamera(new CameraOptions.Builder().bearing(v).build());
            }
        }
    };
    private boolean bearingListenerInstalled = false;

    private OnIndicatorPositionChangedListener mLocationPositionChangeListener = new OnIndicatorPositionChangedListener() {
        @Override
        public void onIndicatorPositionChanged(@NonNull Point point) {
            if (mFollowUserLocation) {
                mMapView.getMapboxMap().setCamera(new CameraOptions.Builder().center(point).build());
                GesturesUtils.getGestures(mMapView).setFocalPoint(mMapView.getMapboxMap().pixelForCoordinate(point));

                // sendUserLocationUpdateEvent(point);
            }
        }
    };

    public LocationComponentManager(RCTMGLMapView rctmglMapView, Context context) {
        mMapView = rctmglMapView;
        mMap = mMapView.getMapboxMap();
        mContext = context;

        mLocationManager = LocationManager.getInstance(context);
    }

    private boolean mShowUserLocation = false;

    private boolean mFollowUserLocation = false;

    private boolean mShowingUserLocation = false;

    public void showUserLocation(boolean showUserLocation) {
        mShowUserLocation = showUserLocation;
        stateChanged();
    }

    public void setFollowUserLocation(boolean followUserLocation) {
        mFollowUserLocation = followUserLocation;
        stateChanged();
    }



    public void setCameraMode(@CameraMode.Mode int cameraMode) {
        mCameraMode = cameraMode;
        stateChanged();

        LocationComponentPlugin locationComponent = LocationComponentUtils.getLocationComponent(mMapView);

        if (mCameraMode == CameraMode.NONE || mCameraMode == CameraMode.TRACKING) {
            locationComponent.removeOnIndicatorBearingChangedListener(mLocationBearingChangedListener);
        } else {
            locationComponent.addOnIndicatorBearingChangedListener(mLocationBearingChangedListener);
        }

        if (mCameraMode == CameraMode.NONE || mCameraMode == CameraMode.NONE_COMPASS || mCameraMode == CameraMode.NONE_GPS) {
            locationComponent.removeOnIndicatorPositionChangedListener(mLocationPositionChangeListener);
        } else {
            locationComponent.addOnIndicatorPositionChangedListener(mLocationPositionChangeListener);
        }
    }

    public void setRenderMode(@RenderMode.Mode int renderMode) {
        mRenderMode = renderMode;
    }
/*
    public void addOnCameraTrackingChangedListener(OnCameraTrackingChangedListener onCameraTrackingChangedListener) {
        // mLocationComponent.addOnCameraTrackingChangedListener(onCameraTrackingChangedListener);
    }*/

    @SuppressLint("MissingPermission")
    private void stateChanged() {
        mLocationComponent.setEnabled((mFollowUserLocation || mShowUserLocation));

        if (mShowingUserLocation != mShowUserLocation) {
            updateShowUserLocation(mShowUserLocation);
        }

        if (mFollowUserLocation) {
            if (!mShowUserLocation) {
               // mLocationComponent.setRenderMode(RenderMode.GPS);
            } else {
                // mLocationComponent.setRenderMode(mRenderMode);
            }
            mLocationComponent.onStart();
        } else {
            // mLocationComponent.setCameraMode(CameraMode.NONE);
        }
        mLocationComponent.setEnabled(mFollowUserLocation || mShowUserLocation);
    }

    public boolean hasLocationComponent() {
        return (mLocationComponent != null);
    }

    public void update(@NonNull Style style) {
        update(mShowUserLocation, style);
    }

    public void update(boolean displayUserLocation, @NonNull Style style) {

        Integer tintColor = mMapView.getSolidColor();

        if (mLocationComponent == null || tintColor != null ) {
            mLocationComponent = LocationComponentUtils.getLocationComponent(mMapView);

            mLocationComponent.setLocationProvider(mLocationManager.getProvider());
            mShowingUserLocation = displayUserLocation;
        }

        updateShowUserLocation(displayUserLocation);
    }

    private void updateShowUserLocation(boolean displayUserLocation) {
        if (mShowingUserLocation != displayUserLocation) {
            applyOptions(displayUserLocation, mLocationComponent);
            mShowingUserLocation = displayUserLocation;
        }
    }

    private void applyOptions(boolean displayUserLocation, LocationComponentPlugin locationComponent) {
        locationComponent.setEnabled(true);
        // Integer tintColor = mMapView.getTintColor();
        if (!displayUserLocation) {
            LocationPuck2D locationPuck = new LocationPuck2D();

            Drawable empty = AppCompatResources.getDrawable(mContext, R.drawable.empty);
            locationPuck.setBearingImage(empty);
            locationPuck.setShadowImage(empty);
            locationPuck.setTopImage(empty);

            locationComponent.setLocationPuck(locationPuck);
        } else /* if (tintColor != null) */ {
            /*
            TypedArray typedArray = mContext.obtainStyledAttributes(
                null, R.styleable.mapbox_MapView, 0,0
            );
            Drawable topImage = typedArray.getDrawable(R.styleable.mapbox_MapView_mapbox_locationComponentLocationPuckLocationPuck2DTopImage);
            */

            LocationPuck2D locationPuck = new LocationPuck2D();

            // ColorFilter filter = new LightingColorFilter(Color.parseColor("#4A90E2"), Color.RED);
            Drawable topImage = AppCompatResources.getDrawable(mContext, R.drawable.mapbox_user_icon);

            /*
            topImage.setColorFilter(filter);

            VectorChildFinder vector = new VectorChildFinder(this, R.drawable.my_vector, imageView);

            VectorDrawableCompat.VFullPath path1 = vector.findPathByName("path1");
            path1.setFillColor(Color.RED);
            */

            locationPuck.setTopImage(topImage);
            Drawable bearingImage = AppCompatResources.getDrawable(mContext, R.drawable.mapbox_user_stroke_icon);
            locationPuck.setBearingImage(bearingImage);
            Drawable shadowImage = AppCompatResources.getDrawable(mContext, R.drawable.mapbox_user_icon_shadow);
            locationPuck.setShadowImage(shadowImage);
            locationComponent.setLocationPuck(locationPuck);
            /*
            locationComponent.setPulsingEnabled(true);
            locationComponent.setPulsingColor(Color.RED);
             */
        }
    }
}

File: android/rctmgl/src/main/java-v10/com/mapbox/rctmgl/components/location/LocationComponentManager.kt
package com.mapbox.rctmgl.components.location

import android.content.Context
import android.graphics.Color
import android.graphics.drawable.Drawable
import android.graphics.drawable.VectorDrawable
import androidx.appcompat.content.res.AppCompatResources
import androidx.lifecycle.Lifecycle
import com.mapbox.maps.plugin.locationcomponent.location
import com.mapbox.rctmgl.components.mapview.RCTMGLMapView
import com.mapbox.maps.plugin.LocationPuck2D
import com.mapbox.maps.plugin.PuckBearingSource
import com.mapbox.maps.plugin.lifecycle.lifecycle
import com.mapbox.rctmgl.R
import com.mapbox.rctmgl.location.LocationManager

/**
 * The LocationComponent on android implements display of user's current location.
 * But viewport seems to be tied to it in the sense that if location is not enabled then it's viewport user tracking is not working.
 * LocationComponentManager attempts to separate that, so that Camera can ask for location tracking independent of display of user current location.
 * And NativeUserLocation can ask for display of user's current location - independent of Camera's user tracking.
 */
class LocationComponentManager(mapView: RCTMGLMapView, context: Context) {
    var mMapView = mapView
    var mContext = context
    private var mState = State(showUserLocation=false, followUserLocation=false, hidden=false, tintColor= null, bearingImage = null, puckBearingSource =null)

    private var mLocationManager: LocationManager = LocationManager.getInstance(context)

    private var mNeedsFullUpdate = true

    private var mLocationManagerStarted = false

    data class State(
        val showUserLocation: Boolean,
        val followUserLocation: Boolean,
        val hidden: Boolean, // in case it isn't native
        val tintColor: Int?, // tint of location puck
        var bearingImage: Drawable?, // bearing image (background)
        var puckBearingSource: PuckBearingSource? // bearing source
    ) {
        val enabled: Boolean
            get() = showUserLocation || followUserLocation
    }

    fun update(newStateCallback: (currentState: State) -> State) {
        val newState = newStateCallback(mState);
        if (newState != mState) {
            applyStateChanges(mState, newState)
            mState = newState
        }
    }

    private fun applyStateChanges(oldState: State, newState: State) {
        mMapView.let {
            val needsFullUpdate = mNeedsFullUpdate

            applyStateChanges(it, oldState, newState,  needsFullUpdate)

            mNeedsFullUpdate = false
        }
    }

    private fun applyStateChanges(map: RCTMGLMapView, oldState: State, newState: State, fullUpdate: Boolean) {
        if (map.getLifecycleState() != Lifecycle.State.STARTED) {
            // In case lifecycle was already stopped, so we're part of shutdown, do not call updateSettings as it'll just restart
            // the loationComponent that will not be stopped. See https://github.com/mapbox/mapbox-maps-android/issues/2017
            if (!newState.enabled) {
                stopLocationManager()
            }
            return
        }
        map.location.updateSettings {
            enabled = newState.enabled

            if (fullUpdate || (newState.hidden != oldState.hidden) || (newState.tintColor != oldState.tintColor) || (newState.bearingImage != oldState.bearingImage)) {
                if (newState.hidden) {
                    var emptyLocationPuck = LocationPuck2D()
                    val empty = AppCompatResources.getDrawable(mContext, R.drawable.empty)
                    emptyLocationPuck.bearingImage = empty
                    emptyLocationPuck.shadowImage = empty
                    emptyLocationPuck.topImage = empty
                    //emptyLocationPuck.opacity = 0.0
                    locationPuck = emptyLocationPuck
                    pulsingEnabled = false
                } else {
                    val mapboxBlueColor = Color.parseColor("#4A90E2")
                    val tintColor = newState.tintColor
                    val defaultLocationPuck = LocationPuck2D()
                    var topImage = AppCompatResources.getDrawable(mContext, R.drawable.mapbox_user_icon)
                    if (tintColor != null) {
                        val drawable = topImage as VectorDrawable?
                        drawable!!.setTint(tintColor)
                        topImage = drawable
                    }
                    defaultLocationPuck.topImage = topImage
                    val defaultBearingImage = AppCompatResources.getDrawable(
                        mContext, R.drawable.mapbox_user_stroke_icon
                    );
                    defaultLocationPuck.bearingImage = newState.bearingImage ?: defaultBearingImage
                    val shadowImage = AppCompatResources.getDrawable(
                        mContext, R.drawable.mapbox_user_icon_shadow
                    )
                    defaultLocationPuck.shadowImage = shadowImage
                    locationPuck = defaultLocationPuck
                    pulsingEnabled = true
                    if (tintColor != null) {
                        pulsingColor = tintColor
                    } else {
                        pulsingColor = mapboxBlueColor
                    }
                }
            }

            if (fullUpdate || newState.enabled != mState.enabled) {
                if (newState.enabled) {
                    startLocationManager()
                    useMapLocationProvider(map)
                } else {
                    stopLocationManager()
                }
            }
        }
    }

    private fun startLocationManager() {
        if (!mLocationManagerStarted) {
            mLocationManager.startCounted()
            mLocationManagerStarted = true
        }
    }

    private fun stopLocationManager() {
        if (mLocationManagerStarted) {
            mLocationManager.stopCounted()
            mLocationManagerStarted = false
        }
    }

    private fun useMapLocationProvider(mapView: RCTMGLMapView) {
        val provider = mapView.location.getLocationProvider()
        if (provider != null) {
            mLocationManager.provider = provider
        }
    }


    fun showNativeUserLocation(showUserLocation: Boolean) {
        update {
            it.copy(showUserLocation = showUserLocation)
        }
    }

    fun setFollowLocation(followUserLocation: Boolean) {
        update {
            it.copy(followUserLocation = followUserLocation)
        }
    }

    fun update() {
        update { it }
    }

    fun onDestroy() {
        stopLocationManager();
    }
}